### PR TITLE
Journal: Autotracking: Cutting-Edge CS Powering Beautiful DX

### PR DESCRIPTION
**Pull request by chriskrycho**
A deep dive into autotracking and rerendering and getters.

- [x] first draft
- [x] red-pen edit on paper
- [x] revise
- [ ] review for correctness (@pzuraq)

### CODE REVIEW

**Comment by chriskrycho** [comment]
@pzuraq if you get 10 minutes today to review this for correctness, that would be *suuuuuper* helpful. 🙏

File: site/journal/2020/Autotracking Elegant DX Via Cutting-Edge CS.md [hunk 1]
```diff
@@ -0,0 +1,548 @@
...
+
+  get remaining() {
+    return this.nameLength - MAX_LENGTH;
+  }
+
+  get showError() {
+    return this.remaining < 0;
+  }
+
+  updateNameTo = (value) => this.myName = value;
```
Comment (pzuraq) [inline site/journal/2020/Autotracking Elegant DX Via Cutting-Edge CS.md:101]: I wonder if using arrow function syntax here might be confusing to some folks. I understand why you're doing it, to follow the Glimmer component example, but it's a bit tangential, and I don't think necessary for this particular example.

Comment (pzuraq) [reply site/journal/2020/Autotracking Elegant DX Via Cutting-Edge CS.md:101]: On the flip side, explaining why you need to change it to be an arrow function later could also be tangential and confusing

File: site/journal/2020/Autotracking Elegant DX Via Cutting-Edge CS.md [hunk 2]
```diff
@@ -0,0 +1,548 @@
...
+const MAX_LENGTH = 10;
+
+class PersonInfo {
+  @tracked name = '';
+}
+```
+
+—it turns into something which acts more like this, where `markAsUsed` says that a property was *read* and `markAsChanged` says it was *set*:
+
+```js
```
Comment (pzuraq) [inline site/journal/2020/Autotracking Elegant DX Via Cutting-Edge CS.md:360]: This example does not syntax highlight for some reason

File: site/journal/2020/Autotracking Elegant DX Via Cutting-Edge CS.md [hunk 3]
```diff
@@ -0,0 +1,548 @@
...
+
+### (3) Recomputing
+
+When rendering templates,[^reactive-again] the runtime sets up what is called a *tracking frame* for each new “computation” in the UI—values, components, helpers, modifiers, etc. A tracking frame is basically just a list of all the tracked properties that called `markAsUsed` while computing any particular value in the template. Since each tracking frame corresponds to a dynamic element of the UI, evaluating the entire UI the first time it is rendered produces a tree of tracking frames which corresponds exactly to the tree of UI components. Critically, though, a tracking frame doesn’t store the *values* of the tracked properties referenced during its computation. Instead, the frame stores only a reference to each property along with the property’s current and previous global clock values.
```
Comment (pzuraq) [inline site/journal/2020/Autotracking Elegant DX Via Cutting-Edge CS.md:500]: This sentence could use a graphic explainer I think, it's difficult to capture but some thing showing the tree would really help drive how the structure that is being setup

Comment (pzuraq) [reply site/journal/2020/Autotracking Elegant DX Via Cutting-Edge CS.md:500]: One option would be to have a tree of tracking frames, and a screengrab of a rendered UI element, and show the tracking frames corresponding to their portion of the UI (perhaps matching numbers/colors/both)

File: site/journal/2020/Autotracking Elegant DX Via Cutting-Edge CS.md [hunk 4]
```diff
@@ -0,0 +1,548 @@
...
+- Re-renders are about as cheap as they possibly can be: all the state computations are simple integer math.
+- Intermediate, “derived” state gets computed on demand when the state it depends on changes—but with normal JavaScript semantics, without extra developer-facing boilerplate or end-user impact on performance.
+- It’s trivial to layer your own caching or memoization on top of these semantics if you need them, but you only pay for what you need.
+- All the “smarts” lives at the very edge of the system, in root state marked with `@tracked` and leaf values computed in reactive contexts like templates.
+
+Hopefully this has give you a good idea how autotracking works in general, and specifically how it simultaneously enables most of our code to be “just JavaScript” *and* gives us a very low-cost reactivity.
+
+:::callout
+
+If you’d like to see some of the details of how these pieces are implemented, check out [the video][walkthrough-video] of my conversation with Ember core team member and Glimmer VM contributor [Chris Garrett][cg] ([@pzuraq](https://github.com/pzuraq/)).
```
Comment (pzuraq) [inline site/journal/2020/Autotracking Elegant DX Via Cutting-Edge CS.md:525]: Any chance you could link to my original autotracking blog post series, or talk from EmberConf? Not looking for exposure or anything, I just think those might be helpful as well 😛

File: site/journal/2020/Autotracking Elegant DX Via Cutting-Edge CS.md [hunk 5]
```diff
@@ -0,0 +1,548 @@
...
+true
+```
+
+In this example, the JavaScript I’ve written evaluates the values directly when logging them. When we use a value in a template in Ember or Glimmer apps, the template engine (the Glimmer VM) evaluates those values. The VM uses a lightweight reactivity system called *autotracking* to track which items in the UI need to be updated in any render. The next step, then, is understanding autotracking.
+
+[^hooks]: This is actually a critical part of how React Hooks work under the hood.
+
+[^closures-classes]: It’s also worth seeing how closures are the [dual](https://en.wikipedia.org/wiki/Duality_(mathematics)) of classes. These two have *the same semantics* as far as an end user is concerned:
+
+    ```js
```
Comment (pzuraq) [inline site/journal/2020/Autotracking Elegant DX Via Cutting-Edge CS.md:281]: This example doesn't appear to highlight correctly

Comment (chriskrycho) [reply site/journal/2020/Autotracking Elegant DX Via Cutting-Edge CS.md:281]: Yeah, it’s highlightjs/highlight.js#2700. I’m debating on the trade off with just removing the private field in favor of a traditional “private” field, vs. leaving it as is and getting syntax highlighting once the upstream issue is fixed. Thoughts?

File: site/journal/2020/Autotracking Elegant DX Via Cutting-Edge CS.md (hunk 3 above)
Comment (chriskrycho) [reply site/journal/2020/Autotracking Elegant DX Via Cutting-Edge CS.md:500]: Heh. I agree. I have a note written in red pen on my draft which says “graphic?” I’ll figure something out. Thanks for calling it out—good confirmation for me.

File: site/journal/2020/Autotracking Elegant DX Via Cutting-Edge CS.md (hunk 4 above)
Comment (chriskrycho) [reply site/journal/2020/Autotracking Elegant DX Via Cutting-Edge CS.md:525]: Yes! In fact, I meant to for the autotracking series but had forgotten about it for the EmberConf video. Great call.

File: site/journal/2020/Autotracking Elegant DX Via Cutting-Edge CS.md (hunk 1 above)
Comment (chriskrycho) [reply site/journal/2020/Autotracking Elegant DX Via Cutting-Edge CS.md:101]: Yeah, I’ve gone back and forth on this. I think I’ll add a footnote calling it out. My thought was that I either need to do it this way *or* I need to explain binding with `@action` or via arrow function and it was easier to go this way given an expected/hoped-for broader audience than just Ember devs. :thinking:

File: site/journal/2020/Autotracking Elegant DX Via Cutting-Edge CS.md [hunk 6]
```diff
@@ -0,0 +1,548 @@
...
+```handlebars
+<div>
+  <input {{on "input" this.updateName}} value={{this.myName}} />
+  <p class='{{if this.showError "error"}}'>
+    ({{this.nameLength}} remaining)
+  </p>
+</div>
+```
+
+There are a handful of interesting features to note about this code’s approach to reactivity. We decorate *one* piece of state, `name`, with `@tracked`, and the rest of the state updates automatically—including the `showError` and `remaining` properties don’t even refer to `name` directly. All of this with a particularly light touch:
```
[the remark this answers is not in the document]
Comment (chriskrycho) [reply site/journal/2020/Autotracking Elegant DX Via Cutting-Edge CS.md:65]: Good catch, thank you.

File: site/journal/2020/Autotracking Elegant DX Via Cutting-Edge CS.md (hunk 5 above)
Comment (chriskrycho) [reply site/journal/2020/Autotracking Elegant DX Via Cutting-Edge CS.md:286]: I went with just using `_` names instead; the point is conveyed just as well that way and I've already clarified that this is not the real implementation. Better to just avoid distracting readers and move on.

**Pull request merged**